spi/control-plane/contract-spi/src/main/java/org/eclipse/edc/connector/controlplane/contract/spi/types/offer/ContractOffer.java

```diff
@@ -103,8 +104,17 @@ public Builder policy(Policy policy) {
 
         public ContractOffer build() {
             Objects.requireNonNull(contractOffer.id, "Id must not be null");
+            var offerId = ContractOfferId.parseId(contractOffer.id)
+                    .orElseThrow(failure -> new IllegalArgumentException(failure.getFailureDetail()));
             Objects.requireNonNull(contractOffer.assetId, "Asset id must not be null");
             Objects.requireNonNull(contractOffer.policy, "Policy must not be null");
+            if (!offerId.assetIdPart().equals(contractOffer.assetId)) {
+                throw new IllegalArgumentException("Asset id %s must be equal to asset id part of offer id %s".formatted(contractOffer.assetId, offerId.assetIdPart()));
+            }
+
+            if (!contractOffer.assetId.equals(contractOffer.policy.getTarget())) {
+                throw new IllegalArgumentException("Policy target %s must be equal to asset id %s".formatted(contractOffer.policy.getTarget(), contractOffer.assetId));
+            }
```

Comment:
generally speaking these builder validations are somewhat dangerous e.g. if in the database we have entities that won't pass this validation there will be un-manageable extensions thrown, plus it makes testing really annoying (as demonstrated by the changelog of this PR)

We have 4 `ContractOffer` new instance creations in the codebase, and 3 of them already take the policy asset id to set the `assetId`, it's just a matter to add a validation before the remaining call (that, looking at the code, it's not really needed).